app/gen-server/lib/homedb/UsersManager.ts

```diff
@@ -0,0 +1,759 @@
+  // unseen user fires off multiple api calls.
+  public async getUserByLoginWithRetry(email: string, options: GetUserOptions = {}): Promise<User|undefined> {
+    try {
+      return await this.getUserByLogin(email, options);
+    } catch (e) {
+      if (e.name === 'QueryFailedError' && e.detail &&
+          e.detail.match(/Key \(email\)=[^ ]+ already exists/)) {
+        // This is a postgres-specific error message. This problem cannot arise in sqlite,
+        // because we have to serialize sqlite transactions in any case to get around a typeorm
+        // limitation.
+        return await this.getUserByLogin(email, options);
+      }
+      throw e;
+    }
+  }
+
+  /**
+   * Find a user by email. Don't create the user if it doesn't already exist.
+   */
+  public async getExistingUserByLogin(
+    email: string,
+    manager?: EntityManager
+  ): Promise<User|undefined> {
+    const normalizedEmail = normalizeEmail(email);
+    return await (manager || this._connection).createQueryBuilder()
+      .select('user')
+      .from(User, 'user')
+      .leftJoinAndSelect('user.logins', 'logins')
+      .where('email = :email', {email: normalizedEmail})
+      .getOne() || undefined;
+  }
+
+  /**
+   *
+   * Fetches a user record based on an email address.  If a user record already
+   * exists linked to the email address supplied, that is the record returned.
+   * Otherwise a fresh record is created, linked to the supplied email address.
+   * The supplied `options` are used when creating a fresh record, or updating
+   * unset/outdated fields of an existing record.
+   *
+   */
+  public async getUserByLogin(email: string, options: GetUserOptions = {}): Promise<User|undefined> {
+    const {manager: transaction, profile, userOptions} = options;
+    const normalizedEmail = normalizeEmail(email);
+    const userByLogin = await this._runInTransaction(transaction, async manager => {
+      let needUpdate = false;
+      const userQuery = manager.createQueryBuilder()
+        .select('user')
+        .from(User, 'user')
+        .leftJoinAndSelect('user.logins', 'logins')
+        .leftJoinAndSelect('user.personalOrg', 'personalOrg')
+        .where('email = :email', {email: normalizedEmail});
+      let user = await userQuery.getOne();
+      let login: Login;
+      if (!user) {
+        user = new User();
+        // Special users do not have first time user set so that they don't get redirected to the
+        // welcome page.
+        user.isFirstTimeUser = !NON_LOGIN_EMAILS.includes(normalizedEmail);
+        login = new Login();
+        login.email = normalizedEmail;
+        login.user = user;
+        needUpdate = true;
+      } else {
+        login = user.logins[0];
+      }
+
+      // Check that user and login records are up to date.
+      if (!user.name) {
+        // Set the user's name if our provider knows it.  Otherwise use their username
```

Comment:
```suggestion
        // Set the user's name if our provider knows it.  Otherwise use their username
```
```suggestion
        // Set the user's name if our provider knows it. Otherwise use their username
```